### PR TITLE
Efficient download

### PR DESCRIPTION
**Pull request by Baapity**
Minimized the time it takes for the lambda-functions to run
* csdapi key usage not secure. just plain text

lambda functions added:
send_request, check_results & their respective states

image
pair coded with @joakimmoller

### CODE REVIEW

File: app/app_stack.py [hunk 1]
```diff
@@ -129,13 +207,44 @@ def __init__(self, scope: Construct, id: str, **kwargs) -> None:
         )
         checkfile_exists_state.when(
             sfn.Condition.number_equals("$.CheckFile.Payload.StatusCode", 404),
-            download_era5_task,
+            send_request_task,
         )
         checkfile_exists_state.otherwise(check_file_fail_state)
+
+        wait_state = sfn.Wait(
+            self, "Wait", time=sfn.WaitTime.duration(Duration.seconds(10))
```
Comment (skymandr) [inline app/app_stack.py:215]: How did you conclude this was a good number? Seems a little on the short side to me.

File: app/download/handler/download_era5.py [hunk 2]
```diff
@@ -1,34 +1,24 @@
...
-    except Exception as err:
-        if "too many" in str(err):
-            raise CDSAPITooManyRequests()
-        if "yet" in str(err):
-            raise CDSAPINotAvailableYet()
-        raise err
+    s3_client = boto3.client("s3")
+    client = Client(progress=False, wait_until_complete=False)
+    result = Result(client, event["reply"])
+    with tempfile.NamedTemporaryFile() as f:
```
Comment (skymandr) [inline app/download/handler/download_era5.py:14]: I don't see that this Lambda has `ephemeral_storage` activated; does this work without it?

**Review by skymandr — APPROVED** [review]
Looks good, but check the storage thing.

File: app/download/handler/download_era5.py (hunk 2 above)
Comment (joakimmoller) [reply app/download/handler/download_era5.py:14]: 512Mb is always available in /tmp

File: app/app_stack.py (hunk 1 above)
Comment (joakimmoller) [reply app/app_stack.py:215]: The standard cdsapi client starts with 1 second and for each try it multiplies with 1.5 until it reaches 120s, it retries 500 times.

Our first try completed after the first wait-state, obviously no load on the service at that time. Something like 10s - 30s should be reasonable.

**Review by joakimmoller — APPROVED** [review]

**Pull request merged**